src/datachain/client/gcs.py

```diff
@@ -24,12 +25,85 @@ class GCSClient(Client):
     protocol = "gs"
 
     @classmethod
-    def create_fs(cls, **kwargs) -> GCSFileSystem:
+    def _try_authenticate(cls, gcreds, method: str) -> bool:
+        """Attempt to authenticate using the specified method.
+
+        Args:
+            gcreds: Google credentials object
+            method: Authentication method to try
+
+        Returns:
+            bool: True if authentication succeeded, False otherwise
+        """
+        from google.auth.exceptions import GoogleAuthError
+
+        try:
+            gcreds.connect(method=method)
+            return True
+        except (GoogleAuthError, ValueError):
+            # Reset credentials if authentication failed (reverts to 'anon' behavior)
+            gcreds.credentials = None
+            return False
+
+    @classmethod
+    def _get_default_credentials(cls, **kwargs) -> dict[str, Any]:
+        """Get default GCS credentials using various authentication methods.
+
+        Returns:
+            dict: Updated kwargs with appropriate token
+        """
+        from gcsfs.core import DEFAULT_PROJECT
+        from gcsfs.credentials import GoogleCredentials
+        from google.auth.compute_engine._metadata import is_on_gce
+
+        def request_callback(*args, **kwargs):
+            from google.auth.exceptions import TransportError
+
+            raise TransportError("Skip metadata check")
+
+        # If credentials provided in env var, use those
         if os.environ.get("DATACHAIN_GCP_CREDENTIALS"):
             kwargs["token"] = json.loads(os.environ["DATACHAIN_GCP_CREDENTIALS"])
-        if kwargs.pop("anon", False):
-            kwargs["token"] = "anon"  # noqa: S105
+            return kwargs
+
+        # If token is provided, use it
+        if "token" in kwargs:
+            return kwargs
+
+        # If anonymous access requested, force anonymous access
+        if kwargs.get("anon"):
+            kwargs["token"] = ANONYMOUS_TOKEN
+            return kwargs
+
+        # Try various authentication methods
+        gcreds = GoogleCredentials(
+            token=ANONYMOUS_TOKEN,
+            project=DEFAULT_PROJECT,
+            access="full_control",
+        )
+
+        # Define authentication methods to try
+        auth_methods = ["google_default", "cache"]
+        if is_on_gce(request_callback):
```

Comment:
@amritghimire what is the logic behind the request_callback? will always raise or only in certain cases?

Reply:
It is to disable the connection to metadata server. It raises exception and the is_on_gce will check for existence of a file to determine if certain file exist. You can look into the code of is_on_gce for more context.

Reply:
why can't it be disabled on the fsspec side? does it mean that metadata server can be ignored, or does it mean that it won't be retrying anymore?

Reply:
It won't be retrying it anymore. 
Also, this is just a proof of concept PR, if we want to go this route or not. Needs refinement if we want to go this way. Created a PR this way so it is easier to view the difference.

Reply:
It is not in fsspec. It is in `from google.auth.compute_engine._metadata import is_on_gce` as you can see from the import above.

Reply:
`fsspec` can use it the same way we do here, quickly check in the cloud auth path (probably that's what Martin was suggesting in the corresponding fsspec discussion, but I don't remember the details by now)

Reply:
btw, why is it considered safe? was there a reason for this retries or it's just a mistake on the Google lib side?

Reply:
Before we go that route, let's see if this works for us or not first internally IMO.

This is a hacky fix for now to be sure.

 @0x2b3bfa0 may have more context on this.  If we can skip the metadata check or not. 

I was wondering is it safe for datachain to skip the metadata check altogether or not? 


So, users on google compute engine could use token to cloud to force this way of communication.

Reply:
I got sidetracked by a lot of things and I lack enough context to respond. Please ping me in a few day if you still need my input. Sorry! 🙈

Reply:
> I was wondering is it safe for datachain to skip the metadata check altogether or not?

answer should be the same as for fsspec to my mind, or google lib

> So, users on google compute engine could use token to cloud to force this way of communication.

they could, but probably won't expect the discrepancy

Reply:
As I mentioned, this is a hacky fix. We can look for proper way to fix this or look for proper way either to suggest it to fsspec or google lib or use a request which doesn't retry for 5 times when it fails for gce check.

We first need to figure out or discuss if we want to implement something like this. @shcheklein cc. @iterative/datachain

Reply:
If we don't want some hacky fix like this here, I don't think we should spend more time to see how to skip the metadata check here.